### PR TITLE
Fixes #23683 - Fix importing os fact

### PR DESCRIPTION
**Pull request by mdellweg**
When a host is assigned to a new operating system by the fact importer, make
sure, that medium and ptabel is available to the new OS.

### CODE REVIEW

**Comment by timogoebel** [comment]
@ares: You mentioned some issues with this in the Redmine issue. I'd vote to merge this. I want to leave the final decision to you, though.

**Review by timogoebel — APPROVED** [review]

**Comment by ares** [comment]
Yeah, for the concern, please see http://projects.theforeman.org/issues/23683#note-2. I think this could lead to unexpected assignments. Perhaps we need to have more granular "Ignore facts for operating system" setting (and other similar settings)?

**Comment by lzap** [comment]
We do have a fact filter now, we can add to the setting description something like "to filter more, use filter". Hmmm. Something better :-)

**Comment by mdellweg** [comment]
I think, the biggest problem with this topic is, that after updating the OS from facts, the host fails further validations, which means, the database is left in an invalid state. Something like this should never happen.
Does the filter mechanism really prevent this?

**Comment by timogoebel** [comment]
I believe this patch makes sense and I'm having a hard time following the reasoning why either a filtering or ignore setting would help. If I understand this correctly, we want the fact importing to happen - but in a correct way.

**Comment by sbernhard** [comment]
Any news?

**Comment by ares** [comment]
I don't want to block the PR but it feels this can lead to unexpected results. Perhaps it would be better to just don't assign the OS if that leads to invalid host and just render a notification to host owner that we received facts but the host couldn't be updated since because of `host.errors.full_messages`

**Comment by tbrisker** [comment]
While in the normal use case this makes sense, please consider the following case:
A host is provisioned from foreman using CentOS as the operating system. J. Sysadmin decides to manually re-provision the host with Fedora. The host finishes the provisioning and reports to puppet. Fedora now gets the CentOS media added to it, and the host still can't reprovision from foreman since the medium defined for it isn't correct still. 
I agree this is somewhat of an edge case, but still worth considering. A different approach may say that we should go the other way around - If the OS changes, set the medium and ptable to match the new OS. 
TBH, I'm not sure there is a good solution for this that will solve all possible use cases, other than splitting the host provisioning configuration from the host's current reported configuration, as we mentioned briefly on discourse during the discussion regarding hostgroups.

**Comment by mdellweg** [comment]
@tbrisker Well that sounds a little bit constructed, but it is the first example i understand to produce problems.
I take from all of this now:
What we do right now is mixing information from different sources. This is generally dangerous and here it leaves the database in an invalid state.
Also, when i press the rebuild button, i would expect the host to be built with the exact same parameters (given, i did not change anything manually) as before. So I'm really not a big fan of updating the "provisioning OS".
I suggest to do the following:
* Stop changing any provisioning data from imported facts.
* When displaying the "actual OS/arch/...", look for the corresponding puppet fact on the fly and fall back to the "provisioning OS/arch/...".

Maybe in the future:
* Show a warning, if they differ.
* Provide an action to fix it.

**Comment by ekohl** [comment]
> Also, when i press the rebuild button, i would expect the host to be built with the exact same parameters (given, i did not change anything manually) as before. So I'm really not a big fan of updating the "provisioning OS".

A counter example would be a Debian install that's upgraded in place. Now the fact changes the OS version from 8 to 9. If you would reprovision, you probably want Debian 9 instead of 8.

This is hard ...

**Comment by mdellweg** [comment]
@ekohl That is, where the warning would be handy.

**Comment by ekohl** [comment]
Yes, but it would change the current behavior. I don't mind the change per se, but it is a breaking change for some users and needs to be in the release notes.

**Comment by timogoebel** [comment]
I think we should keep the current behavior and let users decide via a setting what they want.

**Comment by ekohl** [comment]
We already have `update_environment_from_facts`, `update_subnets_from_facts`. I guess a third wouldn't be that bad but perhaps we need a more generalized solution. An array setting `update_from_facts` which defaults to `['environment', 'operatingsystem', 'subnets']` could be extensible.

**Pull request closed without merging**
